tests/gradients/core/test_pulse_generator_gradient.py

```diff
@@ -0,0 +1,623 @@
+        par_fn_jac = jax.grad(integral_of_polyval)(params[0], t)
+
+        op = qml.evolve(H)(params, t)
+        gens = _one_parameter_generators(op)
+
+        # op has a single parameter in its parameter input list, so the
+        # generators should be a length-1 tuple.
+        assert isinstance(gens, tuple)
+        assert len(gens) == 1
+        gen = gens[0]
+        # The effective generator should have the shape of the parameters and the pulse
+        # together. Here, the parameters are one-dimensional, so we get (num_params, dim, dim)
+        assert gen.shape == (len(params[0]), dim := 2 ** len(term.wires), dim)
+        # The outer product of the rotation parameter Jacobian and the matrix of
+        # the Hamiltonian term yields the Jacobian of the pulse
+        expected = jnp.tensordot(par_fn_jac, -1j * term.matrix(), axes=0)
+        assert qml.math.allclose(gen, expected)
+
+    @pytest.mark.parametrize(
+        "terms",
+        [
+            [X(0), Z(1), X(0) @ Z(1)],
+            [X(0), X(0)],
+            [X(0), Z(1), Y(3)],
+            [
+                Y("a") @ X(3),
+                X(3) @ Z(0),
+                Y("a") @ Z(0),
+            ],
+        ],
+    )
+    @pytest.mark.parametrize("t", ([0.3, 0.4], [-0.1, 0.1]))
+    def test_with_commuting_timedep_terms_ham(self, terms, t):
+        """Test that the generators are correct for a Hamiltonian with multiple
+        commuting time-dependent terms."""
+        import jax
+        import jax.numpy as jnp
+
+        jax.config.update("jax_enable_x64", True)
+
+        num_terms = len(terms)
+        H = qml.math.dot([jnp.polyval for _ in range(num_terms)], terms)
+        params = [jnp.array([0.4, 0.1, 0.2]), jnp.array([0.9, -0.2, 0.5]), jnp.array([-0.5, 0.2])]
+        params = params[:num_terms]
+        # Jacobian functions of the effective rotation parameter (all polyval)
+        par_fn_jac_fn = jax.grad(integral_of_polyval)
+        # Jacobian of the effective rotation parameter for the different terms
+        par_fn_jac = [par_fn_jac_fn(p, t) for p in params]
+
+        op = qml.evolve(H, dense=True, atol=1e-9)(params, t)
+        gens = _one_parameter_generators(op)
+
+        # op has as many parameters in its parameter input list as there are terms,
+        # so generators should be a tuple with as many entries as Hamiltonian terms.
+        assert isinstance(gens, tuple)
+        assert len(gens) == num_terms
+        dim = 2 ** len(H.wires)
+        for gen, term, p, jac in zip(gens, terms, params, par_fn_jac):
+            # The effective generator should have the shape of the parameters and the pulse
+            # together. Here, the parameters are one-dimensional, so we get (num_params, dim, dim)
+            assert gen.shape == (len(p), dim, dim)
+            expected = jnp.tensordot(
+                jac, -1j * expand_matrix(term.matrix(), term.wires, H.wires), axes=0
+            )
+            assert qml.math.allclose(gen, expected)
+
```

Comment:
There is a test for multiple non-commuting constant terms - should there be a similar one for non-commuting time-dependent terms?

Reply:
I was wondering about the same. The problem is that we can not compute the expected value in a second route (i.e. using more analysis) for this case, so we basically would be checking the function against a reimplementation of itself. If it makes sense, I can add it.

Reply:
I guess it's still nice to have in the sense that then if the code gets messed up when someone is doing something in the future, the test will catch it. Not sure if there's really a situation that would break this and not the previous tests though.

Reply:
Good point. I'll just include it :)

Reply:
Done :)